extension/observer/k8sobserver/node_endpoint.go

```diff
@@ -0,0 +1,106 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sobserver // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer/k8sobserver"
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"go.opentelemetry.io/collector/config"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+// convertNodeToEndpoint converts a node instance into a k8s.node observer.Endpoint. It will determine the
+// Target by the first address match of InternalIP, InternalDNS, HostName, ExternalIP, and ExternalDNS in that
+// order.
+func convertNodeToEndpoint(idNamespace string, node *v1.Node) observer.Endpoint {
+	nodeID := observer.EndpointID(fmt.Sprintf("%s/%s-%s", idNamespace, node.Name, node.UID))
+
+	var internalIP, internalDNS, hostname, externalIP, externalDNS string
+
+	for _, addr := range node.Status.Addresses {
+		switch addr.Type {
+		case v1.NodeInternalIP:
+			internalIP = addr.Address
+		case v1.NodeInternalDNS:
+			internalDNS = addr.Address
+		case v1.NodeHostName:
+			hostname = addr.Address
+		case v1.NodeExternalIP:
+			externalIP = addr.Address
+		case v1.NodeExternalDNS:
+			externalDNS = addr.Address
+		}
+	}
+
+	var target string
+	for _, candidate := range []string{internalIP, internalDNS, hostname, externalIP, externalDNS} {
+		if candidate != "" {
+			target = candidate
+			break
+		}
+	}
+
+	// These fields are cleared to prevent excessive endpoint churn/receiver cycling
+	node.ResourceVersion = ""
+	for i := range node.Status.Conditions {
+		node.Status.Conditions[i].LastHeartbeatTime = metav1.Time{}
+		node.Status.Conditions[i].LastTransitionTime = metav1.Time{}
+	}
+
+	var metadata, spec, status map[string]interface{}
+	for _, item := range []struct {
+		src interface{}
+		tgt *map[string]interface{}
+	}{
+		{node.ObjectMeta, &metadata}, {node.Spec, &spec}, {node.Status, &status},
+	} {
+		var jsonMap map[string]interface{}
+		if marshaled, err := json.Marshal(item.src); err == nil {
+			if err := json.Unmarshal(marshaled, &jsonMap); err == nil {
+				configMap := config.NewMap()
+				for k, v := range jsonMap {
+					configMap.Set(k, v)
+				}
+				*(item.tgt) = configMap.ToStringMap()
+			}
+		}
+	}
+
+	nodeDetails := observer.K8sNode{
+		UID:                 string(node.UID),
+		Annotations:         node.Annotations,
+		Labels:              node.Labels,
+		Name:                node.Name,
+		InternalIP:          internalIP,
+		InternalDNS:         internalDNS,
+		Hostname:            hostname,
+		ExternalIP:          externalIP,
+		ExternalDNS:         externalDNS,
+		KubeletEndpointPort: uint16(node.Status.DaemonEndpoints.KubeletEndpoint.Port),
+		Metadata:            metadata,
+		Spec:                spec,
+		Status:              status,
```

Comment:
Why do we need to keep all these fields for Node endpoint while others have only a predefined set fields?

Reply:
I'm sorry I may not understand your question. These fields are for exposing node objects for deployment-specific expressions instead of predefining more limited node info. I personally added them to keep in line with existing Smart Agent functionality that Splunk users have been provided but opted for elevating the kubelet port info and using json formatting over its existing direct node struct access: https://github.com/signalfx/signalfx-agent/blob/main/deployments/k8s/serverless/configmap.yaml#L53

Reply:
I'm just thinking why don't we keep these fields (metadata, spec and status) for pod endpoints but only for node endpoints. I don't see where these fields are used in this PR. I would imagine that end user might be more interested in some nested fields from metadata/spec/status of a pod rather than of a node. But so far I don't remember any issues of asks for such data.

Let me know if I'm missing something

Reply:
They're not used in the observer directly but are available to the user in the receiver creator rules: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6823/files#diff-197f7f6aa6aa4e445182ea1c1ff4394ef2e0f7b266dbc58bd3b5ed6afca4feedR52

I'm not sure why they weren't added to pod/port initially since I wasn't involved in that but they could always be added? One potential downside of them is that they can introduce endpoint churn if chatty event content isn't removed like is being proposed here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6820/files#diff-878c61202da338de920d843bd60c8d2f61c6c3a92f198350cabb2800974751e4R61

Reply:
Yes, the churn is another concerns for this comment. If some other ephemeral fields inside any of these structures are added in future, it can introduce the churn without us noticing. So I'm still wondering why do we need these field? Are they that important for the end users?

Reply:
I think the node spec and status objects have some helpful information but do wonder if promoting it to their own fields as needed is a better way to go than marshaling what the api provides directly. Metadata is largely redundant.

Reply:
Yes, this sounds better to me as well.